osu.Game/Tests/Visual/TestPlayer.cs

```diff
@@ -81,8 +81,22 @@ private void load()
             ScoreProcessor.NewJudgement += r => Results.Add(r);
         }
 
+        public ISkin Skin { get; private set; }
+
+        private TestSkinProvidingContainer rulesetSkinProvider;
+
+        internal void SetSkin(ISkin skin)
```

Comment:
why isn't this in the constructor?

Reply:
I'm still not seeing where this requirement comes from. What causes the existing tests with their existing setup (ie. on `master`) to not work? Shouldn't the fallback order mean they are can just continue to exist as they are without this weirdness?

Reply:
That would need the `CreatePlayer` method and `TestPlayer` constructor signatures to change, which would result in a huge amount of files that need changing.

That's the main reason why I've switched to an internal method / property instead, avoids all the hassle and effort to get it done properly with `constructor`.

Reply:
> I'm still not seeing where this requirement comes from. What causes the existing tests with their existing setup (ie. on `master`) to not work? Shouldn't the fallback order mean they are can just continue to exist as they are without this weirdness?

If the default value of `SkinManager.CurrentSkin` (which is the default skin) provides resources for a certain lookup, that would break the test scene as the `SkinManager` would take priority over whatever is cached on top of `Player`.

Not sure what's the "weirdness" part here, is it the internal `SetSkin` method? could attempt and go the constructor way if you're fine with the number of changes it'll cause in this PR.

Reply:
The weird part is that this whole thing is *adding* complexity. And I guarantee that when even *you* look at this in a week's time you won't understand the order of lookup.

Which seems to be something like Player -> skinmanager.current -> parents??

Reply:
You have a point... The fact that `RulesetSkinProvidingContainer` has `SkinManager` skins added inside of it is a major down of this PR, and I've suggested transforming them at a `SkinManager` level where they are as I feel it may be a better choice here, and would completely get rid of the `TestPlayer.SetSkin` part. (except that any test scene skin would have to surround itself with the transformer, which is at least fine than this complexity I think)

Reply:
As a side note, the reason I've initially added the `SkinManager.DefaultSkin` to the `RulesetSkinProvidingContainer` was to attempt making it less complex to understand quickly how the order of lookup goes. 

In fact, I think `RulesetSkinProvidingContainer` should completely disable the ability to fall back to parents, to make it even easier to understand. Thoughts?

Reply:
If that makes things read better I think it's fine. Would need explicit commenting mentioning the fact, and probably similar commenting in `SkinManager` mentioning that gameplay lookups do not normally hit `SkinManager`.

Reply:
What do you think about the `SetSkin` method, should I leave it as-is for now, or would you highly prefer moving it to constructor no matter the number of file changes required?

Reply:
I still believe it cannot exist. Nor should the constructor change. I'll take a look at this one because I've lost faith in this PR.

Reply:
I've made a change which should allow all test changes to be reverted (to using their DI based lookup) and remove the need for the `SetSkin` method. Could you please revert the necessary test changes and make sure this looks okay to your eyes?

Reply:
Should be done now, the concept of `ISkinSource.AllSources` and using them on `RulesetSkinProvidingContainer` to re-expose all the skins in the hierarchy while transforming legacy ones sounds to be pretty good. No issues other than the two side commits I pushed along with the test changes revert.